core/src/Module/Sql/DataSource/ArrayRawSqlDataSource.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Harmony\Core\Module\Sql\DataSource;
+
+use Harmony\Core\Data\DataSource\DeleteDataSource;
+use Harmony\Core\Data\DataSource\GetDataSource;
+use Harmony\Core\Data\DataSource\PutDataSource;
+use Harmony\Core\Data\Error\DataNotFoundException;
+use Harmony\Core\Data\Error\QueryNotSupportedException;
+use Harmony\Core\Data\Query\AllQuery;
+use Harmony\Core\Data\Query\Composed\ComposedQuery;
+use Harmony\Core\Data\Query\IdsQuery;
+use Harmony\Core\Data\Query\Query;
+use Harmony\Core\Data\Query\VoidQuery;
+use Harmony\Core\Module\Sql\Schema\SqlSchemaInterface;
+use Harmony\Core\Module\Sql\SqlBuilder;
+use InvalidArgumentException;
+
+/**
+ * @see RawSqlProductInteractorsTest
+ *
+ * @implements GetDataSource<mixed[]>
+ * @implements PutDataSource<mixed[]>
+ */
+class ArrayRawSqlDataSource implements
+  GetDataSource,
+  PutDataSource,
+  DeleteDataSource {
+  protected readonly RawSqlDataSource $rawSqlDataSource;
+
+  public function __construct(
+    protected readonly SqlServiceInterface $pdo,
+    protected readonly SqlBuilder $sqlBuilder,
+    protected readonly SqlSchemaInterface $schema,
+  ) {
+    $this->sqlBuilder->setSchema($schema);
+    $this->rawSqlDataSource = new RawSqlDataSource($pdo, $sqlBuilder, $schema);
+  }
+
+  /**
+   * @return mixed[]
+   * @throws DataNotFoundException
+   * @throws QueryNotSupportedException
+   */
+  public function get(Query $query): array {
+    $sql = match (true) {
+      $query instanceof AllQuery => $this->sqlBuilder->selectAll(),
+      $query instanceof ComposedQuery => $this->sqlBuilder->selectAllComposed(
+        $query,
+      ),
+      default => throw new QueryNotSupportedException($query),
+    };
+
+    $items = $this->pdo->findAll($sql->sql(), $sql->params());
+
+    if (empty($items)) {
+      throw new DataNotFoundException();
+    }
+
+    return $items;
+  }
+
+  /**
+   * @param Query         $query
+   * @param object[]|null $entity
+   *
+   * @return mixed[]
+   * @throws QueryNotSupportedException|DataNotFoundException
+   *
+   * @todo           At present this method is not returning the inserted
+   *                 or updated entities.
+   */
+  public function put(Query $query, mixed $entity = null): array {
+    if (!is_array($entity)) {
```

Comment:
is_array is true when we have any type of object. Should we check if the array is a sequential list?
https://www.w3docs.com/snippets/php/how-to-check-if-php-array-is-associative-or-sequential.html#:~:text=To%20check%20if%20an%20array,then%20the%20array%20is%20sequential.

Reply:
Fixed.